serve/tests/unittest/test_sampler.py

```diff
@@ -0,0 +1,327 @@
+        get_sampling_metadata([sampling_param])
+
+    with pytest.raises(ValueError):
+        logit_bias = {1: 10, 3: -10, vocab_size + 100: 2}
+        sampling_param = SamplingParams(logit_bias=logit_bias)
+        get_sampling_metadata([sampling_param])
+
+    with pytest.raises(ValueError):
+        logit_bias = {1: 10, -1: -10}
+        sampling_param = SamplingParams(logit_bias=logit_bias)
+        get_sampling_metadata([sampling_param])
+
+
+def _test_logit_bias():
+    # test single batch
+    batch_size = 1
+    shape = (batch_size, vocab_size)
+    logits = torch.rand(shape, dtype=dtype, device=dev)
+    logit_bias = {1: -1, 3: 1, 2: 2}
+    sampling_param = SamplingParams(logit_bias=logit_bias)
+    sampling_metadata = get_sampling_metadata([sampling_param])
+
+    expected = torch.clone(logits)
+    for idx, val in logit_bias.items():
+        expected[0][idx - 1] += val
+    new_logits = adjust_logits(logits, sampling_metadata, vocab_size)
+    assert torch.allclose(expected, new_logits)
+
+    # test multi-batch
+    batch_size = 3
+    shape = (batch_size, vocab_size)
+    logits = torch.rand(shape, dtype=dtype, device=dev)
+    list_logit_bias = [{1: -1, 3: 1, 2: 2}, {4: 2, 5: 1}, {1: -10}]
+    sampling_params = [
+        SamplingParams(logit_bias=logit_bias) for logit_bias in list_logit_bias
+    ]
+    sampling_metadata = get_sampling_metadata(sampling_params)
+
+    expected = torch.clone(logits)
+    for batch_size in range(batch_size):
+        logit_bias = list_logit_bias[batch_size]
+        for idx, val in logit_bias.items():
+            expected[batch_size][idx - 1] += val
+    new_logits = adjust_logits(logits, sampling_metadata, vocab_size)
+    assert torch.allclose(expected, new_logits)
+
+
+def _test_penalties_checker():
+    get_sampling_metadata([SamplingParams(presence_penalty=-1.0)])
+    get_sampling_metadata([SamplingParams(frequency_penalty=-1.0)])
+    get_sampling_metadata([SamplingParams(repetition_penalty=0.7)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata([SamplingParams(presence_penalty=-2.1)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata([SamplingParams(frequency_penalty=-2.1)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata([SamplingParams(repetition_penalty=-2.1)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata([SamplingParams(presence_penalty=2.1)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata([SamplingParams(frequency_penalty=2.1)])
+
+    with pytest.raises(ValueError):
+        get_sampling_metadata(
+            [
+                SamplingParams(frequency_penalty=1.1),
+                SamplingParams(repetition_penalty=2.1),
+                SamplingParams(presence_penalty=1.1),
+                SamplingParams(presence_penalty=3.1),
+            ]
+        )
+
+
+def _test_penalties():
+    # TODO(vvchernov): Add test for repetition penalty
```

Comment:
@vvchernov, would you add one after this PR? I couldn't find much info about repetition penaltiy in OpenAI spec.